packages/moonstone/VideoPlayer/Video.js

```diff
@@ -0,0 +1,273 @@
+
+		/**
+		 * Video component to use. The default (`'video'`) renders an `HTMLVideoElement`. Custom
+		 * video components must have a similar API structure, exposing the following APIs:
+		 *
+		 * Properties:
+		 * * `currentTime` {Number} - Playback index of the media in seconds
+		 * * `duration` {Number} - Media's entire duration in seconds
+		 * * `error` {Boolean} - `true` if video playback has errored.
+		 * * `loading` {Boolean} - `true` if video playback is loading.
+		 * * `paused` {Boolean} - Playing vs paused state. `true` means the media is paused
+		 * * `playbackRate` {Number} - Current playback rate, as a number
+		 * * `proportionLoaded` {Number} - A value between `0` and `1`
+		 *	representing the proportion of the media that has loaded
+		 * * `proportionPlayed` {Number} - A value between `0` and `1` representing the
+		 *	proportion of the media that has already been shown
+		 *
+		 * Methods:
+		 * * `play()` - play video
+		 * * `pause()` - pause video
+		 * * `load()` - load video
+		 *
+		 * The [`source`]{@link moonstone/VideoPlayer.VideoBase.source} property is passed to
+		 * the video component as a child node.
+		 *
+		 * @type {String | Element | Component}
+		 * @default 'video'
+		 * @public
+		 */
+		mediaComponent: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.element]),
+
+		/**
+		 * Disable automatically playing the video after it has loaded
+		 *
+		 * @type {Boolean}
+		 * @default false
+		 * @public
+		 */
+		noAutoPlay: PropTypes.bool,
+
+		/**
+		 * The video source to be preloaded. Expects a `<source>` node.
+		 *
+		 * @type {String|Node}
+		 * @public
+		 */
+		preloadSource:  PropTypes.node,
+
+		/**
+		 * Called with a reference to the active [Media]{@link ui/Media.Media} component
+		 *
+		 * @type {Function}
+		 * @private
+		 */
+		setMedia: PropTypes.func,
+
+		/**
+		 * The video source to be played.
+		 *
+		 * Any children `<source>` elements will be sent directly to the `mediaComponent` as video
+		 * sources.
+		 *
+		 * See: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/source
+		 *
+		 * @type {String | Node}
+		 * @public
+		 */
+		source: PropTypes.oneOfType([PropTypes.string, PropTypes.node])
+	}
+
+	static defaultProps = {
+		mediaComponent: 'video'
+	}
+
+	componentDidUpdate (prevProps) {
+		const {source, preloadSource} = this.props;
+		const {source: prevSource, preloadSource: prevPreloadSource} = prevProps;
+
+		const key = getKeyFromSource(source);
+		const prevKey = getKeyFromSource(prevSource);
+		const preloadKey = getKeyFromSource(preloadSource);
+		const prevPreloadKey = getKeyFromSource(prevPreloadSource);
+
+		if (this.props.setMedia !== prevProps.setMedia) {
+			this.clearMedia(prevProps);
+			this.setMedia();
+		}
+
+		if (source) {
+			if (key === prevPreloadKey) {
+				// if there's source and it was the preload source
+
+				// if the preloaded video didn't error, notify VideoPlayer it is ready to reset
+				if (!this.video.error) {
```

Comment:
If this is the case, shouldn't we be firing `onReady` when `canplaythrough` event is fired instead of `loadstart`?